sympy/core/basic.py

```diff
@@ -1720,73 +1699,119 @@ def _eval_derivative_n_times(self, s, n):
+        ========
 
-        Pattern as a list of functions:
+        If *pattern* is unspecified, all possible expressions are transformed.
 
-        >>> sin(x).rewrite([sin, ], exp)
-        -I*(exp(I*x) - exp(-I*x))/2
+        >>> from sympy import cos, sin, exp, I
+        >>> from sympy.abc import x
+        >>> expr = cos(x) + I*sin(x)
+        >>> expr.rewrite(exp)
+        exp(I*x)
+
+        Pattern can be a type or an iterable of types.
+
+        >>> expr.rewrite(sin, exp)
+        exp(I*x)/2 + cos(x) - exp(-I*x)/2
+        >>> expr.rewrite([cos,], exp)
+        exp(I*x)/2 + I*sin(x) + exp(-I*x)/2
+        >>> expr.rewrite([cos, sin], exp)
+        exp(I*x)
+
+        Rewriting behavior can be implemented by defining ``_eval_rewrite()``
+        method.
+
+        >>> from sympy import Expr, sqrt, pi
+        >>> class MySin(Expr):
+        ...     def _eval_rewrite(self, rule, args, **hints):
+        ...         x, = args
+        ...         if rule == cos:
+        ...             return cos(pi/2 - x, evaluate=False)
+        ...         if rule == sqrt:
+        ...             return sqrt(1 - cos(x)**2)
+        >>> MySin(MySin(x)).rewrite(cos)
+        cos(-cos(-x + pi/2) + pi/2)
+        >>> MySin(x).rewrite(sqrt)
+        sqrt(1 - cos(x)**2)
 
         """
         if not args:
             return self
+
+        hints.update(deep=deep)
+
+        pattern = args[:-1]
+        rule = args[-1]
+
+        # support old design
+        if isinstance(rule, str):
+            method = "_eval_rewrite_as_%s" % rule
+        elif hasattr(rule, "__name__"):
+            # rule is class or function
+            clsname = rule.__name__
+            method = "_eval_rewrite_as_%s" % clsname
         else:
-            pattern = args[:-1]
-            if isinstance(args[-1], str):
-                rule = '_eval_rewrite_as_' + args[-1]
-            else:
-                # rewrite arg is usually a class but can also be a
-                # singleton (e.g. GoldenRatio) so we check
-                # __name__ or __class__.__name__
-                clsname = getattr(args[-1], "__name__", None)
-                if clsname is None:
-                    clsname = args[-1].__class__.__name__
-                rule = '_eval_rewrite_as_' + clsname
+            clsname = rule.__class__.__name__
+            method = "_eval_rewrite_as_%s" % clsname
 
+        if pattern:
+            if iterable(pattern[0]):
+                pattern = pattern[0]
+            pattern = tuple(p for p in pattern if self.has(p))
             if not pattern:
-                return self._eval_rewrite(None, rule, **hints)
+                return self
+        else:
+            pattern = None
+
+        def _rewrite(expr, pattern, rule, method, **hints):
+            deep = hints.pop('deep', True)
+            if deep:
+                args = [_rewrite(a, pattern, rule, method, **hints)
+                        if isinstance(a, Basic) else a for a in expr.args]
```

Comment:
Is the Basic check here needed? It shouldn't be necessary: all args should be Basic. There was a problem with this in stats before but I think it got fixed.

It's better to not to have an if like this because it's just designed to allow bugs that should really be fixed.

Reply:
Classes in stats module still has `None` in `args` so I cannot change this in this PR.

Reply:
Which classes? I really think that we should fix things like that rather than have workarounds in core Basic methods. The workarounds just facilitate broken subclasses. It's much better if those subclasses blow up as soon as they're written.

Reply:
Test failures suggest that `Moment` and `CentralMoment` in `sympy.stats.symbolic_probability` module have `None` in args.

This workaround is in master branch as well. It will be nice if we fix the stats module and remove this in the future, but I think it should not be a blocker for this PR.

Reply:
Test failures suggest that `Moment` and `CentralMoment` in `sympy.stats.symbolic_probability` module have `None` in args.

Reply:
This workaround is in master branch as well because of this reason. Perhaps I can re-define rewrite method for these classes and remove the workaround for `Basic`?

Reply:
@oscarbenjamin 

I removed the Basic check from `Basic.rewrite()`.

Reply:
It's not actually that difficult to fix them e.g.:
```diff
diff --git a/sympy/stats/symbolic_probability.py b/sympy/stats/symbolic_probability.py
index c74207f2a9..6057e5d5d5 100644
--- a/sympy/stats/symbolic_probability.py
+++ b/sympy/stats/symbolic_probability.py
@@ -603,8 +603,9 @@ def __new__(cls, X, n, c=0, condition=None, **kwargs):
         n = _sympify(n)
         c = _sympify(c)
         if condition is not None:
-            condition = _sympify(condition)
-        return Expr.__new__(cls, X, n, c, condition)
+            return Expr.__new__(cls, X, n, c, _sympify(condition))
+        else:
+            return Expr.__new__(cls, X, n, c)
 
     def doit(self, **hints):
         if not is_random(self.args[0]):
```

Reply:
Done.